// vscode/vscode/src/vs/workbench/services/configurationResolver/common/configurationResolverSchema.js
import { localizeWithPath } from '../../../../nls.js';
const _moduleId = "vs/workbench/services/configurationResolver/common/configurationResolverSchema";
const idDescription = ( localizeWithPath(
    _moduleId,
    0,
    "The input's id is used to associate an input with a variable of the form ${input:id}."
));
const typeDescription = ( localizeWithPath(_moduleId, 1, "The type of user input prompt to use."));
const descriptionDescription = ( localizeWithPath(
    _moduleId,
    2,
    "The description is shown when the user is prompted for input."
));
const defaultDescription = ( localizeWithPath(_moduleId, 3, "The default value for the input."));
const inputsSchema = {
    definitions: {
        inputs: {
            type: 'array',
            description: ( localizeWithPath(
                _moduleId,
                4,
                'User inputs. Used for defining user input prompts, such as free string input or a choice from several options.'
            )),
            items: {
                oneOf: [
                    {
                        type: 'object',
                        required: ['id', 'type', 'description'],
                        additionalProperties: false,
                        properties: {
                            id: {
                                type: 'string',
                                description: idDescription
                            },
                            type: {
                                type: 'string',
                                description: typeDescription,
                                enum: ['promptString'],
                                enumDescriptions: [
                                    ( localizeWithPath(
                                        _moduleId,
                                        5,
                                        "The 'promptString' type opens an input box to ask the user for input."
                                    )),
                                ]
                            },
                            description: {
                                type: 'string',
                                description: descriptionDescription
                            },
                            default: {
                                type: 'string',
                                description: defaultDescription
                            },
                            password: {
                                type: 'boolean',
                                description: ( localizeWithPath(
                                    _moduleId,
                                    6,
                                    "Controls if a password input is shown. Password input hides the typed text."
                                )),
                            },
                        }
                    },
                    {
                        type: 'object',
                        required: ['id', 'type', 'description', 'options'],
                        additionalProperties: false,
                        properties: {
                            id: {
                                type: 'string',
                                description: idDescription
                            },
                            type: {
                                type: 'string',
                                description: typeDescription,
                                enum: ['pickString'],
                                enumDescriptions: [
                                    ( localizeWithPath(_moduleId, 7, "The 'pickString' type shows a selection list.")),
                                ]
                            },
                            description: {
                                type: 'string',
                                description: descriptionDescription
                            },
                            default: {
                                type: 'string',
                                description: defaultDescription
                            },
                            options: {
                                type: 'array',
                                description: ( localizeWithPath(
                                    _moduleId,
                                    8,
                                    "An array of strings that defines the options for a quick pick."
                                )),
                                items: {
                                    oneOf: [
                                        {
                                            type: 'string'
                                        },
                                        {
                                            type: 'object',
                                            required: ['value'],
                                            additionalProperties: false,
                                            properties: {
                                                label: {
                                                    type: 'string',
                                                    description: ( localizeWithPath(_moduleId, 9, "Label for the option."))
                                                },
                                                value: {
                                                    type: 'string',
                                                    description: ( localizeWithPath(_moduleId, 10, "Value for the option."))
                                                }
                                            }
                                        }
                                    ]
                                }
                            }
                        }
                    },
                    {
                        type: 'object',
                        required: ['id', 'type', 'command'],
                        additionalProperties: false,
                        properties: {
                            id: {
                                type: 'string',
                                description: idDescription
                            },
                            type: {
                                type: 'string',
                                description: typeDescription,
                                enum: ['command'],
                                enumDescriptions: [
                                    ( localizeWithPath(_moduleId, 11, "The 'command' type executes a command.")),
                                ]
                            },
                            command: {
                                type: 'string',
                                description: ( localizeWithPath(_moduleId, 12, "The command to execute for this input variable."))
                            },
                            args: {
                                oneOf: [
                                    {
                                        type: 'object',
                                        description: ( localizeWithPath(_moduleId, 13, "Optional arguments passed to the command."))
                                    },
                                    {
                                        type: 'array',
                                        description: ( localizeWithPath(_moduleId, 13, "Optional arguments passed to the command."))
                                    },
                                    {
                                        type: 'string',
                                        description: ( localizeWithPath(_moduleId, 13, "Optional arguments passed to the command."))
                                    }
                                ]
                            }
                        }
                    }
                ]
            }
        }
    }
};
export { inputsSchema };
