// vscode/vscode/src/vs/platform/contextkey/common/contextkeys.js
import { isWeb, isMacintosh, isLinux, isWindows, isIOS, isMobile } from '../../../base/common/platform.js';
import { localizeWithPath } from '../../../nls.js';
import { RawContextKey } from './contextkey.js';
const _moduleId = "vs/platform/contextkey/common/contextkeys";
const IsMacContext = ( (new RawContextKey('isMac', isMacintosh, ( localizeWithPath(_moduleId, 0, "Whether the operating system is macOS")))));
const IsLinuxContext = ( (new RawContextKey('isLinux', isLinux, ( localizeWithPath(_moduleId, 1, "Whether the operating system is Linux")))));
const IsWindowsContext = ( (new RawContextKey('isWindows', isWindows, ( localizeWithPath(_moduleId, 2, "Whether the operating system is Windows")))));
const IsWebContext = ( (new RawContextKey('isWeb', isWeb, ( localizeWithPath(_moduleId, 3, "Whether the platform is a web browser")))));
const IsMacNativeContext = ( (new RawContextKey('isMacNative', isMacintosh && !isWeb, ( localizeWithPath(
 _moduleId,
 4,
 "Whether the operating system is macOS on a non-browser platform"
)))));
const IsIOSContext = ( (new RawContextKey('isIOS', isIOS, ( localizeWithPath(_moduleId, 5, "Whether the operating system is iOS")))));
const IsMobileContext = ( (new RawContextKey('isMobile', isMobile, ( localizeWithPath(_moduleId, 6, "Whether the platform is a mobile web browser")))));
const IsDevelopmentContext = ( (new RawContextKey('isDevelopment', false, true)));
const ProductQualityContext = ( (new RawContextKey('productQualityType', '', ( localizeWithPath(_moduleId, 7, "Quality type of VS Code")))));
const InputFocusedContextKey = 'inputFocus';
const InputFocusedContext = ( (new RawContextKey(InputFocusedContextKey, false, ( localizeWithPath(_moduleId, 8, "Whether keyboard focus is inside an input box")))));
export { InputFocusedContext, InputFocusedContextKey, IsDevelopmentContext, IsIOSContext, IsLinuxContext, IsMacContext, IsMacNativeContext, IsMobileContext, IsWebContext, IsWindowsContext, ProductQualityContext };
